lnst/Recipes/ENRT/ConfigMixins/LongLivedConnectionsMixin.py

```diff
@@ -0,0 +1,230 @@
+    This mixin adds support for long-lived connections.
+
+    Based on `long_lived_conns` parameter, it will create a number of
+    long-lived connections between the hosts. Receiver is a server, while
+    generator is a client.
+
+    Only long lived connections IPs are handled by this mixin.
+    Therefore, if your test requires perf IPs, it should be
+    configured by parent's test_wide_configuration() method.
+    IPs used for long-lived client and server are added based on
+    long_lived_conns_per_ip parameter, which defines size of
+    addressable space for connections. Sinde L4 can address up
+    to 65535 ports, it's obviously limited by that.
+
+    Connections are not equally distributed among clients.
+    The first client (and others) will get long_lived_conns_per_ip
+    connections, while the last one will get the remaining connections.
+
+    Don't forget to set appropriate NO_FILES ulimit (if needed).
+    See LongLivedServer/LongLivedClient for more details.
+    """
+
+    long_lived_conns = IntParam(mandatory=True)
+    long_lived_conns_port = IntParam(default=20000)
+    long_lived_conns_per_ip = IntParam(default=20000)
+
+    def _generate_ip_endpoints(self, config):
+        return [
+            ip_endpoint_pairs(
+                config,
+                (self.matched.host1.eth0, self.matched.host2.eth0),
+                combination_func=zip,
+            )
+        ]
+
+    def _filter_ip_endpoints(self, config, slicer):
+        ips = []
+
+        for parallel_endpoints in self._generate_ip_endpoints(config):
+            ipv4 = [
+                endpoint_pair
+                for endpoint_pair in parallel_endpoints
+                if isinstance(endpoint_pair.first.address, Ip4Address)
+            ]
+            ipv6 = [
+                endpoint_pair
+                for endpoint_pair in parallel_endpoints
+                if isinstance(endpoint_pair.first.address, Ip6Address)
+            ]
+
+            ips.append(ipv4[slicer])
+            ips.append(ipv6[slicer])
+
+        return ips
+
+    def generate_perf_endpoints(self, config):
+        return self._filter_ip_endpoints(
+            config, slice(0, 1)  # only the first IP is used for perf
+        )
+
+    def generate_long_lived_conns_endpoints(self, config):
+        return self._filter_ip_endpoints(
+            config,
+            slice(1, None),  # all IPs except the first one which is used for perf
+        )
+
+    @property
+    def servers_count(self):
+        return ceil(self.params.long_lived_conns / self.params.long_lived_conns_per_ip)
+
+    def calculate_client_connections(self, client_id):
+        client_id += 1  # 0-based index
+        if client_id < self.servers_count:
+            return self.params.long_lived_conns_per_ip
+
+        # remaining connections are handled by last client
+        return (
+            self.params.long_lived_conns
+            - (self.servers_count - 1) * self.params.long_lived_conns_per_ip
```

Comment:
isn't this just `self.params.long_lived_conns % self.servers_count`? as in modulo (remainder)

Reply:
Not really, there are couple of edge cases comming from `%` returns 0 if `long_lived_conns` is dividable by `servers_count`, so it would need to look like:


```python
    def calculate_client_connections(self, client_id):
        client_id += 1  # 0-based index
        remaining_conns = self.params.long_lived_conns % self.servers_count

        if client_id < self.servers_count or not remaining_conns:
            if self.params.long_lived_conns < self.params.long_lived_conns_per_ip:
                return self.params.long_lived_conns

            return self.params.long_lived_conns_per_ip

        # remaining connections are handled by last client
        return remaining_conns
```

But i've refactored my implementation a bit, hope it's more readable, wdyt?

Reply:
thanks, looks better.